File: Task4/Kotlin/Task4_stream.kt
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*

fun createFlow(array: List<Int>): Flow<Int> = flow {
    array.forEach { item ->
        emit(item)
    }
}

suspend fun <T, R> streamMap(
    stream: Flow<T>,
    callback: suspend (T) -> R,
    signal: Job
): List<R> {
    val results = mutableListOf<R>()
    try {
        stream.collect { value ->
            if (!signal.isActive) {
                throw CancellationException("Aborted")
            }
            results.add(callback(value))
        }
    } catch (e: CancellationException) {
        throw e
    }
    return results
}

fun demoFunc() = runBlocking {
    val controller = Job()

    // case 1
    val numbers = listOf(1, 2, 3, 4, 5)
    val promiseTriple: suspend (Int) -> Int = { num -> num * 3 }

    val stream = createFlow(numbers)

    try {
        val results1 = streamMap(stream, promiseTriple, controller)
        println("Case 1 results: $results1")
    } catch (err: CancellationException) {
        println("Case 1 was aborted")
    } catch (err: Exception) {
        println("Error: ${err.message}")
    }

    controller.cancel()

    // case 2
    val numbers2 = listOf(10, 20, 30)
    val promiseSquare: suspend (Int) -> Int = { num -> num * num }

    val stream2 = createFlow(numbers2)

    try {
        val results2 = streamMap(stream2, promiseSquare, controller)
        println("Case 2 results: $results2")
    } catch (err: CancellationException) {
        println("Case 2 was aborted")
    } catch (err: Exception) {
        println("Error: ${err.message}")
    }
}

fun main() {
    demoFunc()
}

File: Task4/JavaScript/Task4_stream.js
async function streamMap(stream, callback, signal) {
    const reader = stream.getReader();
    const results = [];

    try {
        while (true) {
            if (signal.aborted) {
                throw new DOMException("Aborted", "AbortError");
            }

            const { done, value } = await reader.read();

            if (done) break;

            results.push(await callback(value));
        }
    } finally {
        reader.releaseLock();
    }

    return results;
}

function createReadableStream(array) {
    return new ReadableStream({
        start(controller) {
            array.forEach((item) => controller.enqueue(item));
            controller.close();
        },
    });
}

async function demoFunc() {
    const controller = new AbortController();
    const { signal } = controller;

    const numbers = [1, 2, 3, 4, 5];
    const promiseTriple = (num) => Promise.resolve(num * 3);

    const stream = createReadableStream(numbers);

    try {
        const results1 = await streamMap(stream, promiseTriple, signal);
        console.log("Case 1 results:", results1);
    } catch (err) {
        if (err.name === "AbortError") {
            console.error("Case 1 was aborted");
        } else {
            console.error("Error:", err);
        }
    }

    controller.abort();

    const numbers2 = [10, 20, 30];
    const promiseSquare = (num) => Promise.resolve(num * num);

    const stream2 = createReadableStream(numbers2);

    try {
        const results2 = await streamMap(stream2, promiseSquare, signal);
        console.log("Case 2 results:", results2);
    } catch (err) {
        if (err.name === "AbortError") {
            console.error("Case 2 was aborted");
        } else {
            console.error("Error:", err);
        }
    }
}

demoFunc()
    .then(() => console.log("demoFunc completed successfully"))
    .catch((err) => console.error("Error in demoFunc:", err));
